File: Templates/BIT.java

public class BIT{

    private int maxSize;
    private long[] tree;

    public BIT(){
        maxSize = 100000;
        tree = new long[maxSize];
    }

    public BIT(int maxSize){
        this.maxSize = maxSize;
        tree = new long[this.maxSize];
    }

    public void update(int i, long val){
        while(i < maxSize){
            tree[i] += val;
            i += (i & -i);
        }
    }

    public long sum(int i){
        long res = 0;
        while(i > 0){
            res += tree[i];
            i -= (i & -i);
        }
        return res;
    }
}

File: Templates/BIT.py
class BIT:
    def __init__(self, nums):
        self.nums = nums
        self.tree = [0] * (len(nums) + 1)
        for i in range(len(nums)):
            self._update(i, nums[i])
    
    def _update(self, i, val):
        i += 1
        while i < len(self.tree):
            self.tree[i] += val
            i += i & (-i)
    
    def update(self, i, val):
        diff = val - self.nums[i]
        self.nums[i] = val
        self._update(i, diff)
    
    def _prefix_sum(self, i):
        i += 1
        res = 0
        while i > 0:
            res += self.tree[i]
            i -= i & (-i)
        return res
    
    def get_sum(self, i, j):
        return self._prefix_sum(j) - self._prefix_sum(i-1)


    def __str__(self):
        return str(self.tree)
